### PR TITLE
Adding GlobalMaxPooling 1D, 2D, 3D

### PR DESCRIPTION
**Pull request by Shashi456**
I've added the max operation to operators, since it wasnt defined earlier.

### CODE REVIEW

File: Sources/DeepLearning/Layer.swift [hunk 1]
```diff
@@ -1009,6 +1009,63 @@ public struct AvgPool2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    ///   - input: The input to the layer.
+    ///   - context: The contextual information for the layer application, e.g. the current learning
+    ///     phase.
+    /// - Returns: The output.
+    @differentiable
+    public func applied(to input: Tensor<Scalar>, in _: Context) -> Tensor<Scalar> {
+        return input.max(alongAxes: [1, 2, 3]).reshaped(to: [input.shape[0], input.shape[4]])
+    }
+}
+
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Remove extra line.

**Comment by Shashi456** [comment]
@dan-zheng I'm working on the test cases now, I will have a seperate PR for them, there are a lot of layers which do not have test cases and I will make a list of the same in #77, it would be easier to track them that way.

File: Sources/DeepLearning/Operators.swift [hunk 2]
```diff
@@ -44,6 +44,7 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
                                     epsilon: epsilon)
         return (value, { v in
             let mean = self.mean(alongAxes: axis)
+            let max = self.max(alongAxes: axis)
```
Comment (tanmayb123) [inline Sources/DeepLearning/Operators.swift]: This constant isn't used anywhere in this vjp - is it necessary?

Comment (Shashi456) [reply Sources/DeepLearning/Operators.swift]: Oh I probably got the idea wrong, i will remove it.

**Comment by Shashi456** [comment]
@rxwei  can this be merged?

**Comment by rxwei** [comment]
Could you pull from master? I wanted to trigger a rebuild.

**Comment by Shashi456** [comment]
@rxwei the docker summary fails because it says the max function is not differentiable, is there any way correct that?

**Comment by tanmayb123** [comment]
We could either add a VJP for the max op to the standard library, or just use the actual "max pooling" operation inside the global max pooling layers (set the pool size to the width & height of the input).

**Comment by Shashi456** [comment]
@tanmayb123 I was trying to do that when i initially added the max function to the batch normalized vjp, i just put it in the wrong place. I think we should rather put in the max function because it is actually used in multiple places, might come in as a handy operation.

**Comment by Shashi456** [comment]
@tanmayb123 any idea about how i could go about adding the vjp for the max function

**Comment by Shashi456** [comment]
@dan-zheng Could you guide me maybe, as to how i can add a max vjp function to the operators file?

File: Sources/DeepLearning/Layer.swift [hunk 3]
```diff
@@ -1009,6 +1009,63 @@ public struct AvgPool2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameters:
+    ///   - input: The input to the layer.
+    ///   - context: The contextual information for the layer application, e.g. the current learning
+    ///     phase.
+    /// - Returns: The output.
+    @differentiable
+    public func applied(to input: Tensor<Scalar>, in _: Context) -> Tensor<Scalar> {
+        return input.max(alongAxes: [1, 2, 3]).reshaped(to: [input.shape[0], input.shape[4]])
```
Comment (jekbradbury) [inline Sources/DeepLearning/Layer.swift]:
```suggestion
        return input.max(squeezingAxes: [1, 2, 3])
```
Similarly for the other instances.

**Comment by Shashi456** [comment]
@rxwei could you tell me how i could go about defining a vjp for the max function? 
I think the vjp is the way to go since we'd need a max function for hinge loss as well.

**Comment by rxwei** [comment]
I'm sorry for the delayed response! Recently the team has been busy with releasing v0.3 and getting some important core pieces like RNNs into the library. I'll definitely help you with defining derivatives for `min` and `max` later this week!

**Comment by rxwei** [comment]
In the meantime, it would be useful to explore our [custom differentiation tutorial](https://www.tensorflow.org/swift/tutorials/custom_differentiation), and see how other derivatives are defined, e.g. [`mean(alongAxes:)`](https://github.com/apple/swift/blob/0d9cb2fee195beac93b23dc627ab181b071d9ff6/stdlib/public/TensorFlow/Gradients.swift#L588). Feel free to post any questions you have.

**Comment by Shashi456** [comment]
@rxwei do you mind triggering a build on this? Thanks to @eaplatanios for adding the max and min vjps. With this, we are done with all the pooling layers.

**Comment by Shashi456** [comment]
@rxwei could you take a quick glance and trigger the build once more? This does pass locally now. 

Also is there any way to keep updating the toolchain directly, instead of continuously downloading nightlies xD.

**Comment by rxwei** [comment]
> Also is there any way to keep updating the toolchain directly, instead of continuously downloading nightlies xD.

You can always build a toolchain from the tensorflow branch on apple/swift using the `utils/build-toolchain-tensorflow` script. However, it often takes quite a while.

**Comment by Shashi456** [comment]
@eaplatanios sorry to bother you with this. But could you tell me if the max function call i wrote for this is wrong? The error that pops is `function not differentiable`

**Comment by eaplatanios** [comment]
@Shashi456 Could you post the full error trace? I can clone and take a look in a bit.

**Comment by Shashi456** [comment]
```
[17/36] Compiling TensorFlow Pooling.swift
/swift-apis/Sources/TensorFlow/Layers/Pooling.swift:375:22: error: expression is not differentiable
        return input.max(squeezingAxes: [1, 2])
                     ^
/swift-apis/Sources/TensorFlow/Layers/Pooling.swift:375:22: note: cannot differentiate functions that have not been marked '@differentiable' and that are defined in other files
        return input.max(squeezingAxes: [1, 2])
                     ^
/swift-apis/Sources/TensorFlow/Layers/Pooling.swift:359:22: error: expression is not differentiable
        return input.max(squeezingAxes: 1)
                     ^
/swift-apis/Sources/TensorFlow/Layers/Pooling.swift:359:22: note: cannot differentiate functions that have not been marked '@differentiable' and that are defined in other files
        return input.max(squeezingAxes: 1)
                     ^
/swift-apis/Sources/TensorFlow/Layers/Pooling.swift:391:22: error: expression is not differentiable
        return input.max(squeezingAxes: [1, 2, 3])
                     ^
/swift-apis/Sources/TensorFlow/Layers/Pooling.swift:391:22: note: cannot differentiate functions that have not been marked '@differentiable' and that are defined in other files
return input.max(squeezingAxes: [1, 2, 3])
```

**Comment by eaplatanios** [comment]
I see. The `Tensor.max(squeezingAxes:)` function is not currently differentiable when the axes are int arrays. I'll make a PR to fix that now.

**Comment by eaplatanios** [comment]
Done in #159.

**Comment by rxwei** [comment]
#159 has been merged. Could you do a pull?

**Comment by Shashi456** [comment]
@rxwei done. pretty happy that finally done with pooling layers xD

File: Tests/TensorFlowTests/LayerTests.swift [hunk 4]
```diff
@@ -124,6 +124,30 @@ final class LayerTests: XCTestCase {
         XCTAssertEqual(output, expected)
     }
 
+    func testGlobalMaxPool1D() {
+        let layer = GlobalMaxPool1D<Float>()
```
Comment (rxwei) [inline Tests/TensorFlowTests/LayerTests.swift]: Could you make sure tests compile locally? This does not compile because the type was defined as `GlobalMaxPooling1D`, not `GlobalMaxPool1D`.

**Comment by Shashi456** [comment]
@rxwei I'll take care of both the PRs, in a while. Sorry!

**Comment by Shashi456** [comment]
@rxwei built and tested with nightly. This PR passes as is. The build now takes substantially more time though somehow xD.

I'm not sure about the test since the linux error you've mentioned popped up, but i rectified the error you mentioned about `GlobalMaxPooling1D`.

**Comment by Shashi456** [comment]
@eaplatanios could you trigger a build on this?

Thank you.

**Comment by Shashi456** [comment]
@rxwei could you trigger a build on this?

**Comment by Shashi456** [comment]
@rxwei The errors aren't from these additions. I think this can be merged,

**Comment by Shashi456** [comment]
@rxwei I think this is ready to be merged.
